Src/PCompiler/CompilerCore/Backend/Java/NameManager.cs

```diff
@@ -77,7 +116,39 @@ protected override string ComputeNameForDecl(IPDecl decl)
                 name = "TMP_" + name.Substring(1);
             }
 
+            // Handle the case where a declaration happens to match the name of the top level
+            // classname (which, as it isn't strictly a declaration in the grammar, doesn't
+            // pass through here)
+            if (name == TopLevelCName)
+            {
+                string cname = decl.GetType().ToString().Split(".").Last();
+                name = $"{name}_{cname}";
+            }
+
             return UniquifyName(name);
         }
+
+        /// <summary>
+        /// Produces the class name for the foreign function bridge class for a given machine.
+        ///
+        /// In the C# code generator, partial classes are used to automatically weave together
+        /// the machine class and its foreign functions.  For instance, a machine called Client
+        /// that relies on foreign functions would have the latter implemented in a partial class
+        /// also called Client.  The C# compiler would then merge the two together into a final
+        /// class definition.
+        ///
+        /// Java does not have this language feature, unfortunately, so we need a different approach:
+        /// we need foreign function writers to implement foreign functions as static methods
+        /// in a class whose name is derived from the monitor class.  The naming convention is
+        /// specified by the return value of this function.
+        ///
+        /// <see url="https://en.wikipedia.org/wiki/Bridge_pattern"/>
+        /// </summary>
+        /// <param name="machineName"></param>
+        /// <returns></returns>
+        public string FFIBridgeForMachine(string machineName)
```

Comment:
The foreign functions can be global as well. What do you do then?

Reply:
Is there a separate file for `machine` level and global functions?

Reply:
Good point - I ought to emit a `GlobalFunctions` like the C# compiler does.